### PR TITLE
Created a client for the wpcom media api

### PR DESCRIPTION
**Pull request by jameslnewell**
#### Changes proposed in this Pull Request

Since `wpcom` is deprecated and to avoid adding large dependencies I created a new client for the `wpcom` media api using `wpcom-request-proxy.

The client will continue to be expanded and refined as part of the `media-library` work e.g. querying, improved error handling etc.

#### Testing instructions

Since the client is a very thin abstraction on top of the HTTP requests I didn't see much value in attempting to write unit tests with the API mocked. Instead I wrote a bunch of stories showing the API in use.

Run `npm run media-library:storybook:start` and play with the stories to see the API working.

Fixes #

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -144,6 +144,7 @@
 		"@automattic/webpack-inline-constant-exports-plugin": "file:./packages/webpack-inline-constant-exports-plugin",
 		"@babel/cli": "7.8.3",
 		"@babel/core": "7.8.3",
+		"@jameslnewell/react-promise": "^3.3.0",
```
Comment (jsnajdr) [inline package.json]: This is not a `devDependency`, is it? It should be defined in `packages/media-library/package.json` as `dependency`.

File: packages/media-library/src/api/fetch.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+/**
+ * External dependencies
+ */
+import proxy from 'wpcom-proxy-request';
+
+let isAuthorised = false;
+
+const authorise = () => {
+	return new Promise( ( resolve, reject ) => {
+		// TODO: if there is an in-progress authorisation, don't try and authorise, instead wait untill it is resolved
```
Comment (jsnajdr) [inline packages/media-library/src/api/fetch.ts]: We use the "authorize" spelling almost all the time 🙂

**Comment by jsnajdr** [comment]
`wpcom-request-proxy` is something like a low level driver that provides a specific implementation of sending REST requests. It doesn't work in Node.js, and it doesn't work in the Desktop app either. For these, we use the `wpcom-xhr-request` driver.

File: packages/media-library/src/api/fetch.ts (hunk 2 above)
Comment (jameslnewell) [reply packages/media-library/src/api/fetch.ts]: Sorry, hard habit to get out of spelling it correctly 😉🤣

**Comment by jameslnewell** [comment]
👍@jsnajdr Thanks for the explanation.

I originally considered supporting NodeJS and started to implement `fetch.ts` and `fetch.browser.ts` until I realised the current build setup in the repository doesn't seem to support `*.browser.ts` files written in Typescript (I think we'd have to switch compiling with `@babel/preset-typescript`).

I stopped pursuing the idea because I think its likely to be pretty rare that the `media-library` will ever be present on a page when it is SSR'd so we can just load the data after the component mounts. However, I wasn't considering an Electron app!

Why doesn't an iframe work in the desktop app? What security model does it have setup?

I think I can address this in a future by having a `ProxyClient` and `XHRClient` that implement the same `Client` interface. That's probably better than the `*.browser.ts` approach since the `wpcom-xhr-request` will likely require some secrets.

**Comment by jsnajdr** [comment]
> Why doesn't an iframe work in the desktop app? What security model does it have setup?

The Desktop app uses OAuth to authenticate to _wordpress.com_, receives a OAuth token and uses it as a part of the `Authorization` header on every REST request. There are no cookies involved.

The iframe proxy works only in a browser and uses your _wordpress.com_ cookies for authentication. To get _wordpress.com_ cookies, you need to POST a form to _wordpress.com/log-in_ as a first party, and then get redirected back to the domain that loads the iframe.

The Desktop app only ever loads web content from the local Electron server (running at `127.0.0.1:41050`) and never loads any other pages from web. It has no way to get _wordpress.com_ cookies.

To summarize, desktop and `wpcom-xhr-request` use OAuth for authentication, browser and `wpcom-request-proxy` use cookies.

File: packages/media-library/src/api/fetch.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
+/**
```
Comment (tyxla) [inline packages/media-library/src/api/fetch.ts]: I wonder how you feel about the prospect of decoupling the data layer from `lib/wpcom`, then abstracting it as a separate package and using that package for performing the requests? After all, I'd expect that the media library itself shouldn't have to reinvent the wheel as it comes to data needs - instead, it should step on something already built. What do you think?

**Review by tyxla — COMMENTED** [review]
First of all, nice work with this 💯 , and thank you for working on it 👏 

Since we're starting this package from scratch, I was wondering what's the plan for it to reuse the existing pieces of the media library, and where the existing code and this package will meet. 

In my head, I was more thinking about doing some work on the existing media library code, to prepare it for being standalone, for example:

* Reduxify all the stores
* Isolate data layer and any necessary data abstractions into packages
* Start moving existing media library components to the package
* Improve the API for external consumers

What are your thoughts about that?

**Comment by jameslnewell** [comment]
My plan was to build and iterate on it in parallel to the code that already exists in order to avoid rework whilst working on the UI. I was planning to go back and make the necessary changes once everything is usable in isolation. 

Other than the "when" to update the existing code it sounds like a similar plan.

File: packages/media-library/src/api/fetch.ts (hunk 3 above)
Comment (jameslnewell) [reply packages/media-library/src/api/fetch.ts]: What do you define as the data layer? It feels like that's what the `Client` is?

Comment (tyxla) [reply packages/media-library/src/api/fetch.ts]: I was talking about the [data layer](https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer) that much of Calypso state already uses.

**Comment by tyxla** [comment]
I think we can close this one as we're no longer planning to package the media library at this point.
